traffic_control/clients/python/to_access/__init__.py

```diff
@@ -0,0 +1,576 @@
+
+.. option:: --to_host HOST
+
+	The :abbr:`FQDN (Fully Qualified Domain Name)` and optionally the port and scheme of the Traffic
+	Ops server. This will override :envvar:`TO_URL`. The format is the same as for :envvar:`TO_URL`.
+	(Default: uses the value of :envvar:`TO_URL`)
+
+.. option:: --to_password PASSWORD
+
+	The password to use when authenticating to Traffic Ops. Overrides :envvar:`TO_PASSWORD`.
+	(Default: uses the value of :envvar:`TO_PASSWORD`)
+
+.. option:: --to_user USERNAME
+
+	The username to use when connecting to Traffic Ops. Overrides :envvar:`TO_USER`. (Default: uses
+	the value of :envvar:`TO_USER`)
+
+Environment Variables
+---------------------
+If defined, :program:`toaccess` scripts will use these environment variables to define their
+connection to and authentication with the Traffic Ops server. Typically, setting these is easier
+than using the long options :option:`--to_host`, :option:`--to_user`, and :option:`--to_password` on
+every invocation.
+
+.. envvar:: TO_PASSWORD
+
+	Will be used to authenticate the user defined by either :option:`--to_user` or :envvar:`TO_USER`.
+
+.. envvar:: TO_URL
+
+	The :abbr:`FQDN (Fully Qualified Domain Name)` of the Traffic Ops server to which the script
+	will connect. The format of this should be :file:`[{http or https}://]{hostname}[:{port}]`. Note
+	that this may optionally start with ``http://`` or ``https://`` (case insensitive), but
+	typically this is unnecessary. Also notice that the port number may be specified, though again
+	this isn't usually required. All :program:`toaccess` scripts will assume that port 443 should be
+	used unless otherwise specified. They will further assume that the protocol is HTTPS unless
+	:envvar:`TO_URL` (or :option:`--to_host`) starts with ``http://``, in which case the default port
+	will also be set to 80 unless otherwise specified in the URL.
+
+.. envvar:: TO_USER
+
+	The name of the user as whom to connect to the Traffic Ops server.
+
+Module Reference
+================
+
+"""
+import json
+import logging
+import os
+import sys
+import typing
+
+import requests
+
+from trafficops.restapi import LoginError, OperationError, InvalidJSONError
+from trafficops.tosession import TOSession
+
+l = logging.getLogger()
+l.disabled = True
+logging.basicConfig(level=logging.CRITICAL+1)
+
+#: The full path to a file used to store the user's Mojolicious athentication cookie (currently unused)
+COOKIEFILE = "" #os.path.expanduser(os.path.join("~", ".to-auth.cookie"))
+
+def setCookie(cookie:str):
+	"""
+	Writes the passed cookie to the :data:`COOKIEFILE` file for later use.
+
+	.. warning::
+		Currently this function is never used, and it depends on the value of :data:`COOKIEFILE`,
+		which at the moment is being set to an empty string.
+	"""
+	with open(COOKIEFILE, "w") as f:
+		f.write(cookie)
+
+def output(r:requests.Response, pretty:bool, reqHeader:bool, respHeader:bool, reqPayload:bool,
+           indent:typing.Union[int, str] = '\t'):
```

Comment:
No, tabs are being used consistently for indentation. These spaces are meant to align the `indent` argument with the beginning of the function argument list. Indentation here would be improper, because this is one function definition, and should be kept at a single indentation level.

Reply:
I don't think it would be improper. If your concern is that the argument list should line up, you can break the line after the opening `(`:
```
def output(
    r:requests.Response, pretty:bool, reqHeader:bool, respHeader:bool, reqPayload:bool, 
    indent:typing.Union[int, str] = '\t'):
```
I also think it would be easier to read with a space between the param name and its type like so: `foo: bool` rather than `foo:bool`.

Reply:
I have mixed feelings about that. What I've been doing is `arg:type = default`. I found it to be the happiest middle ground. `arg: type = default` looks too much like I'm setting a type to a value, and I think we can all agree `arg:type=default` is too cluttered.

Reply:
Well, the cases where the parameters have defaults are much less common than the cases where the parameters do not have defaults. We should aim to improve readability of the common case. Since we heavily use Go in TC, people should be accustomed to seeing `<name> <type> = <value>`, like `var i int = 1`. But I care much less about that than I do about mixing tabs and spaces in the same python file. That is going to turn into a giant mess real quick. It would be better to just choose tabs or choose spaces, none of this "indentation vs alignment" stuff. You can align code to make it more readable by using a standard method of indentation. We also don't want to be in the business of reviewing code and making sure there are actually tabs or spaces where you expect, since you can't see that stuff when reviewing without highlighting and running your mouse over the whitespace. The more mixing there is, the more likely a `TabError` is going to happen.

Reply:
> Assuming we use <arg>: <type> formatting, I propose adopting `<arg>: type = <default>` as a standard.

+1

> Indentation cannot be relied upon to align with text on a preceding line. I don't like spreading a single statement across multiple levels of indentation in a language where indentation has syntactic meaning.

I imagine that most places where you'd even want to align text, any extra indentation used for alignment doesn't really matter in terms of the language's syntax. Some places that come to mind are dict/list/tuple formatting:
```
my_dict = {
    "one": "two",
    "three": "four",
}
my_list = [
    1,
    2,
    3,
]
my_tuple = (
     1,
     2,
)
```
and function args:
```
def my_cool_func(one, two, three, 
                 four):
    do_stuff()
```
You can safely indent and align using "indentation" within a single statement like that. Sure indentation has syntactic meaning, but not within `()[]{}` at least. Sticking to 4-space indentations would alleviate this mixing of tabs and spaces since they can be used for indentation as well as alignment.

Reply:
But indentation and alignment are different things, which is why I don't like using only one character to represent them both. It would mean things like

```python3
          some_var = 3
some_other_var = 4
```
could throw a `SyntaxError`. Not that I actually like the idea of formatting things like that, it's just the first example that came to mind.

Reply:
As far as type hint formatting: done.

Reply:
> But indentation and alignment are different things, which is why I don't like using only one character to represent them both.

The nice thing with just using spaces for both is that what you see is what you get. There's no issues with different editors displaying tabs differently, one space is always rendered as one space. For reviewing purposes this makes reviewing indentation and alignment a lot easier, since you can always tell if things are indented improperly or misaligned. Tabs are ambiguous and depend on the editor you're viewing the code in, so things like using tabs for alignment could be rendered properly in one editor but be completed misaligned in another. I don't really want to have to check every single indentation or alignment just to make sure that tabs are only being used for indentation and spaces are only being used for alignment. I would much prefer for them both to be done with spaces, so that I can easily notice things that are out of place.

Reply:
That's funny, because "what you see is what you get" is exactly what I like about tabs. One tab is one level of indentation, I don't have to stop and mentally parse a line to figure out if it's indented or just continued from the previous line and aligned, and I don't have to make sure that function x is using the same number of spaces as function y for indentation. Well, I guess I do, but the "spaces" are much bigger and their number stays much smaller even as the code's complexity grows.

but spaces vs tabs aside, I don't really feel very strongly about not using

```python.
def func(
    arg: type = default, ...
    more_args: type = default, ...) -> return_type:
```
instead of the way it is. Personally I think it's ugly, but I wouldn't call it downright unreadable, so if it really bothers you to have it the way it currently is then I'll change it. But not immediately, working on something atm, can't guarantee I'll get to it today.

Reply:
I wasn't saying the arg list _should_ be formatting like that, I was just using that as an example of how to align text by just using "indentation" (be it tabs or spaces).

> I don't have to stop and mentally parse a line to figure out if it's indented or just continued from the previous line and aligned

You don't have to do that with the all-spaces approach either. You can't easily tell the difference between a tab and N spaces? Basically I could indent a file with tabs, as well as 8-spaces, and you wouldn't be able to easily tell the difference if your editor displayed tabs as 8 spaces.

Reply:
Another good reason to use tabs is because they're used nearly everywhere else in the codebase. Go code is mandated to use `go fmt` which means all of our Go code uses tab-based indentation, our PerlTidy spec (that I don't think anyone actually uses) specifies the use of tab-based indentation, all of our Java code appears at a glance to use tabs (I think maybe that's also the Eclipse defalut?), and the style guide for our reStructuredText mandates tabs. That last one is of particular importance, the python client (actually all Python code of which I'm aware within the repository) is documented using `sphinx-autodoc`. That means the docstrings are technically part of the documentation, and need to conform to the doc standards. It's usually not a big deal because indentation within a single docstring is rare, but for example in the TOSession object there's a large docstring with several examples nested in `.. code-block` directives which require indentation syntactically.

Reply:
https://stackoverflow.blog/2017/06/15/developers-use-spaces-make-money-use-tabs/ -- nuff said.

In all seriousness though, I mainly just prefer spaces in Python because it's the standard due to PEP8 (even though I don't agree everything in it should be followed religiously). You will see it in virtually every other big open source Python project. It's not done because it's better than using tabs, it's done because it keeps the broader community consistent. If you're a Python dev and have your editor set to 4-space indentations, in most cases you wouldn't have to change your editor settings when switching between projects. Depending on the editor, this isn't really a big deal anyways, since editors should be able to determine what type of indentation is being used in a file and adapt accordingly.

I think it would be fine to stick to tabs consistently in our Python code. What worries me is mixing tabs and spaces in a file (e.g. for indentation vs alignment), because I think that can get messy very easily and would be better to just avoid that class of bug altogether by sticking to one or the other. I don't want to get bit by that bug, but if we can reasonably depend on the linter to catch those kinds of things at PR review time, I think I wouldn't mind allowing tabs and spaces on the same line for indentation and alignment, respectively.

However, I strongly believe we should try to stick to _at least_ the PEP8-defined naming schemes in all of the project's Python code. Go and Java code shouldn't use snake_case, just like C/C++ and Python code shouldn't use camelCase. Standard language naming schemes should always be followed IMO.

Reply:
Pylint can tell the difference between indentation and alignment regardless of whether tabs or spaces are used or even if they're mixed. The configuration needs to be set, but once it is the linter *will* be able to catch that at review time.

I've switched to snake_case naming because it is used predominantly in the client code, as well as the `requests` library on which the whole thing is based. I still and will always object to doing anything for the reason that it conforms to PEP8 because I think that's a dangerous precedent to set.